drivers/hmis/lib/hmis_data_cleanup/util.rb

```diff
@@ -155,6 +173,41 @@ def self.delete_duplicate_exit_records!
       GrdaWarehouse::Hud::Exit.where(id: ids_to_delete).update_all(DateDeleted: Time.current, source_hash: nil)
     end
 
+    # This cleanup routine was added because duplicate EnrollmentIDs were leading to uniqueness violation constraints
+    # when Client records were merged.
+    #
+    # 1) Hard-delete deleted enrollments where there is any NON-deleted Enrollment that shares an EnrollmentID
+    # 2) Hard-delete deleted all but 1 enrollment where there is a set of deleted Enrollments that share an EnrollmentID
+    def self.hard_delete_duplicate_deleted_enrollments!
+      data_source_id = GrdaWarehouse::DataSource.hmis.first.id
+      result = GrdaWarehouse::Hud::Enrollment.with_deleted.
+        where(data_source_id: data_source_id).
+        group(:EnrollmentID, :ProjectID).
+        having('count(*) >1').
+        select('"Enrollment"."EnrollmentID", array_agg("Enrollment"."id") as enrollment_ids')
+
+      enrollments_by_id = Hmis::Hud::Enrollment.with_deleted.
+        where(id: result.map(&:enrollment_ids).flatten).index_by(&:id)
+
+      enrollments_to_delete = []
+      result.each do |res|
+        # processing duplicates for res.EnrollmentID
+        enrollments = res.enrollment_ids.map { |id| enrollments_by_id[id] }
+        raise 'not found' if enrollments.any?(&:nil?)
+
+        # if any are NOT deleted, hard-delete all the deleted ones
+        if enrollments.any? { |e| e.DateDeleted.nil? }
+          enrollments_to_delete.push(*enrollments.select { |e| e.DateDeleted.present? })
+        else
+          # otherwise they are all deleted, so keep 1 of them
+          to_keep = enrollments.max_by(&:DateUpdated)
+          enrollments_to_delete.push(*enrollments.reject { |e| e.id == to_keep.id })
+        end
+      end
+
+      GrdaWarehouse::Hud::Enrollment.with_deleted.where(id: enrollments_to_delete.map(&:id)).delete_all
```

Comment:
this is inefficient but the only way I could figure out how to hard-delete without destroying. `enrollments_to_delete.each(&:delete)` doesn't work.

Reply:
This is just "delete from Enrollment where id in x,yz" right? That should be fast (much faster than `each(&:delete)`).

It is packing like 10k ids into one sql statement though but that would also be fixed by batching it. 

I also wonder if we should log the ids?
```suggestion
      Rails.logger.info "Deleting duplicate ids: #{enrollments_to_delete.inspect}"
      GrdaWarehouse::Hud::Enrollment.with_deleted.where(id: enrollments_to_delete.map(&:id)).delete_all
```

Reply:
yeah, added logging